pilot/cmd/pilot-discovery/server/server_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server_test
+
+import (
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"strings"
+
+	"github.com/golang/protobuf/ptypes"
+	"istio.io/istio/pilot/cmd/pilot-discovery/server"
+	"istio.io/istio/pilot/proxy/envoy"
+)
+
+type env struct {
+	stop   chan struct{}
+	fsRoot string
+}
+
+func (e *env) setup() (int, error) {
+	e.fsRoot = createTempDir()
+	e.stop = make(chan struct{})
+
+	port, err := getFreePort()
+	if err != nil {
+		return 0, err
+	}
+
+	// Create a test pilot discovery service configured to watch the tempDir.
+	args := server.PilotArgs{
+		Namespace: "testing",
+		DiscoveryOptions: envoy.DiscoveryServiceOptions{
+			Port:            port,
+			EnableCaching:   true,
+			EnableProfiling: true,
+		},
+		Mesh: server.MeshArgs{
+			MixerAddress:    "istio-mixer.istio-system:9091",
+			RdsRefreshDelay: ptypes.DurationProto(10 * time.Millisecond),
+		},
+		Config: server.ConfigArgs{
+			FileDir: e.fsRoot,
+		},
+		Service: server.ServiceArgs{
+			// Using the Mock service registry, which provides the hello and world services.
+			Registries: []string{string(server.MockRegistry)},
+		},
+	}
+
+	// Create and setup the controller.
+	s, err := server.NewServer(args, e.stop)
+	if err != nil {
+		return 0, err
+	}
+
+	// Start the server.
+	s.Run()
+	return port, nil
+}
+
+func (e *env) teardown() {
+	close(e.stop)
+
+	// Remove the temp dir.
+	os.RemoveAll(e.fsRoot)
+}
+
+// TODO(nmittler): getFreePort is copied from https://github.com/phayes/freeport
+func getFreePort() (int, error) {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		return 0, err
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return 0, err
+	}
+	defer l.Close()
+	return l.Addr().(*net.TCPAddr).Port, nil
+}
+
+func createTempDir() string {
+	// Make the temporary directory
+	dir, _ := ioutil.TempDir("/tmp/", "monitor")
+	_ = os.MkdirAll(dir, os.ModeDir|os.ModePerm)
+	return dir
+}
+
+// TestListServices verifies that the mock services are available on the Pilot discovery service
+func TestListServices(t *testing.T) {
+	e := env{}
+	port, err := e.setup()
+	if err != nil {
+		t.Error(err)
+	}
+	defer e.teardown()
+
+	// TODO(nmittler): Is there a better way to avoid flakes?
```

Comment:
retries with linear falloff?

Reply:
Do we have an example elsewhere where we solve this sort of problem?

Reply:
Add a `/health` endpoint to server and busy-poll w/timeout until its ready? We could use the same path for readiness probes in production.

Reply:
@ayj that sounds like a good approach.  Is the `/health` endpoint a standard thing in istio? Could you point me to an example of where we do something similar?

Reply:
We don't have a standard yet. It's on the roadmap for production readiness. @jmuk has a PR open for mixer liveliness / readiness probes (see https://github.com/istio/istio/pull/2002). See https://github.com/istio/istio/issues/1467 for the tracking issue for pilot readiness probes.

Reply:
@ayj Thanks for the link! I've updated the TODO to reference the ongoing work on readiness.

I think we should revisit once a final solution lands.